server/controller/src/main/java/ai/starwhale/mlops/datastore/wal/WalLocalFileManager.java

```diff
@@ -0,0 +1,236 @@
+            this.path = info.path;
+            this.maxEntryId = info.maxEntryId;
+            this.size = info.size;
+        }
+
+        @Getter
+        private long index;
+
+        @Getter
+        private Path path;
+
+        @Getter
+        private long maxEntryId;
+
+        @Getter
+        private int size;
+    }
+
+    private final TreeMap<Long, LogFileInfo> logFileMap = new TreeMap<>();
+
+    public WalLocalFileManager(Path cacheDir, int walMaxFileSize, long suggestedFirstIndex) throws IOException {
+        Files.createDirectories(cacheDir);
+        this.cacheDir = cacheDir;
+        this.walMaxFileSize = walMaxFileSize;
+        try (var stream = Files.list(this.cacheDir)) {
+            this.indexes = WalManager.parseIndexes(
+                    stream.map(path -> path.getFileName().toString()).filter(path -> path.startsWith("wal.log.")));
+        }
+        if (this.indexes.isEmpty() || suggestedFirstIndex - 1 > this.indexes.get(this.indexes.size() - 1)) {
+            this.logFileIndex = suggestedFirstIndex;
+        } else {
+            this.logFileIndex = this.indexes.get(this.indexes.size() - 1);
+        }
+    }
+
+    public long getFirstIndex() {
+        return this.indexes.isEmpty() ? -1 : this.indexes.get(0);
+    }
+
+    public LogFileInfo getLogFileInfo(long index) {
+        var info = this.logFileMap.get(index);
+        return info == null ? null : new LogFileInfo(info);
+    }
+
+    public List<LogFileInfo> getAllLogFileInfo() {
+        return this.logFileMap.values().stream().map(LogFileInfo::new).collect(Collectors.toList());
+    }
+
+    public LogFileInfo getCurrentLogFileInfo() {
+        return this.getLogFileInfo(this.logFileIndex);
+    }
+
+    public int write(List<WalEntry> entries) throws IOException {
+        if (this.outputStream == null) {
+            var path = this.getLocalFilePath(this.logFileIndex);
+            if (!Files.exists(path)) {
+                this.initLogFile();
+            } else {
+                var ch = FileChannel.open(path, StandardOpenOption.WRITE, StandardOpenOption.APPEND);
+                ch.truncate(this.currentFileSize);
+                this.outputStream = Channels.newOutputStream(ch);
+            }
+        }
+        int ret = entries.size();
+        int sizeTotal = 0;
+        try {
+            for (int i = 0; i < entries.size(); ++i) {
+                var entry = entries.get(i);
+                var entrySize = CodedOutputStream.computeMessageSizeNoTag(entry);
+                if (WalManager.HEADER.length + entrySize > this.walMaxFileSize) {
+                    log.error("data loss: discard unexpected huge entry. "
+                                    + "size={} table={} schema={} records count={}",
+                            entry.getSerializedSize(),
+                            entry.getTableName(),
+                            entry.getTableSchema(),
+                            entry.getRecordsCount());
+                }
```

Comment:
There will be a dead loop if the first entry is too large without `continue`

Reply:
Done